### ISSUE TITLE
Cigna Explanations Of Benefits render as gibberish

### ISSUE THREAD

**Issue by InPermutation** [mozilla/pdf.js#6397]
I don't wish to supply an example, for patient privacy reasons, but I can attempt to help debug.

Here's what the top right of the document looks like in Firefox 40.0.2 (Windows 64-bit):

![gibberish](http://www.jkrall.net/pdfjs%20gibberish.PNG)

Here's what the same document looks like in Chrome's PDF viewer:

![Customer service...](http://www.jkrall.net/chrome%20pdf%20viewer.PNG)

Windows `Reader` handles it fine, too:

![Customer service...](http://www.jkrall.net/TWINUI.png)
# Document Properties

```
File name: document.pdf
File size: 371 KB (379,875 bytes)
Title: -
Author: TargetStream Technologies
Subject: -
Keywords: -
Creation Date: -
Modification Date: -
Creator: TargetStream Technologies StreamEDP
PDF Producer: TargetStream StreamEDP for Connecticut General Life Insurance Company, Inc.
PDF Version:  1.4
Page Count: 4 
```

What other information would be helpful?

**Comment by InPermutation** [comment]
Weirdly, it looks good in Inspector:

![perfectly good markup](http://www.jkrall.net/Inspector.PNG)

**Comment by yurydelendik** [comment]
Can you also turn on CSS Warning or Errors in the inspector to see if there are some @font-face errors? Can you create a simple example/test PDF without confidential data?

**Comment by InPermutation** [comment]
CSS Warnings/Errors:

```
Unknown pseudo-class or pseudo-element 'selection'.  Ruleset ignored due to bad selector. viewer.css:59:13
Unknown pseudo-class or pseudo-element 'fullscreen'.  Ruleset ignored due to bad selector. viewer.css:110:21
Unknown pseudo-class or pseudo-element 'fullscreen'.  Ruleset ignored due to bad selector. viewer.css:198:37
Unknown pseudo-class or pseudo-element 'fullscreen'.  Ruleset ignored due to bad selector. viewer.css:213:21
Unknown pseudo-class or pseudo-element 'fullscreen'.  Ruleset ignored due to bad selector. viewer.css:221:21
Unknown pseudo-class or pseudo-element 'selection'.  Ruleset ignored due to bad selector. viewer.css:1315:2
PDF ef85d4919760cd5f6a859427de578103 [1.4 TargetStream StreamEDP for Connecticut General Life Insurance Company, Inc. / TargetStream Technologies StreamEDP] (PDF.js: 1.1.114) viewer.js:6397:1
```

I will see what I can do about redacting confidential data from the PDF.

**Comment by InPermutation** [comment]
Here's some of the filter information from the PDF, if that helps at all:

```
PS> grep -ia '/filter' .\document.pdf
<</Filter/FlateDecode/Length 3304>>stream
<</Filter/FlateDecode/Subtype/Form/Type/XObject/Length 226/BBox[0 0 612 -792]/Resources<</Font<</MyriadPro 9 0 R/MyriadProBold 12 0 R>>>>>>stream
<</BitsPerComponent 1/Decode[1 0]/Filter/FlateDecode/Subtype/Image/Type/XObject/Length 1246/ImageMask true/Height 225/Width 680/ColorSpace/DeviceGray>>stream
<</Filter/FlateDecode/Length 2174>>stream
<</Filter/FlateDecode/Length 2221>>stream
<</BitsPerComponent 1/Decode[1 0]/Filter/FlateDecode/Subtype/Image/Type/XObject/Length 35222/ImageMask true/Height 1016/Width 1800/ColorSpace/DeviceGray>>stream
<</Filter/FlateDecode/Length 1520>>stream
<</Filter/FlateDecode/Length 216>>stream
<</Filter/FlateDecode/Length1 100252/Length 66533>>stream
<</Filter/FlateDecode/Length 1796>>stream
<</Filter/FlateDecode/Length1 79308/Length 50994>>stream
<</Filter/FlateDecode/Length 1534>>stream
<</Filter/FlateDecode/Length 1534>>stream
<</Filter/FlateDecode/Length1 79364/Length 51837>>stream
<</Length 177 /Filter/FlateDecode>>
<</Length 129 /Filter/FlateDecode>>
<</Length 153 /Filter/FlateDecode>>
<</Length 118 /Filter/FlateDecode>>
<</Length 176 /Filter/FlateDecode>>
<</Length 195 /Filter/FlateDecode>>
<</Length 107 /Filter/FlateDecode>>
<</Length 170 /Filter/FlateDecode>>
<</Length 152 /Filter/FlateDecode>>
<</Length 205 /Filter/FlateDecode>>
<</Length 186 /Filter/FlateDecode>>
<</Length 133 /Filter/FlateDecode>>
<</Length 196 /Filter/FlateDecode>>
<</Length 168 /Filter/FlateDecode>>
<</Length 182 /Filter/FlateDecode>>
<</Length 120 /Filter/FlateDecode>>
<</Length 181 /Filter/FlateDecode>>
<</Length 129 /Filter/FlateDecode>>
<</Length 172 /Filter/FlateDecode>>
<</Length 117 /Filter/FlateDecode>>
<</Length 81 /Filter/FlateDecode>>
<</Length 168 /Filter/FlateDecode>>
<</Length 116 /Filter/FlateDecode>>
<</Length 176 /Filter/FlateDecode>>
<</Length 166 /Filter/FlateDecode>>
<</Length 164 /Filter/FlateDecode>>
<</Length 98 /Filter/FlateDecode>>
<</Length 174 /Filter/FlateDecode>>
<</Length 98 /Filter/FlateDecode>>
<</Length 188 /Filter/FlateDecode>>
<</Length 150 /Filter/FlateDecode>>
<</Length 180 /Filter/FlateDecode>>
<</Length 195 /Filter/FlateDecode>>
<</Length 170 /Filter/FlateDecode>>
<</Length 128 /Filter/FlateDecode>>
<</Length 174 /Filter/FlateDecode>>
<</Length 197 /Filter/FlateDecode>>
<</Length 16 /Filter/FlateDecode>>
<</Length 168 /Filter/FlateDecode>>
<</Length 121 /Filter/FlateDecode>>
<</Length 232 /Filter/FlateDecode>>
<</Length 1204 /Filter/FlateDecode>>
<</Filter/FlateDecode/Length 1534>>stream
<</Filter/FlateDecode/Length1 80696/Length 52456>>stream
<</Filter/FlateDecode/Length 1534>>stream
<</Filter/FlateDecode/Length1 81436/Length 52869>>stream
```

**Comment by InPermutation** [comment]
OK, hopefully I redacted this appropriately - here is the 2nd page Glossary of one of my EOBs: https://dl.dropboxusercontent.com/u/2987530/Cigna%20glossary.pdf

(If I didn't redact it appropriately, well, it's for last year's flu shot. Hopefully that's not particularly sensitive information.)

**Comment by Snuffleupagus** [comment]
The main issue (total gibberish) was fixed by PR #6270, which will be included in Firefox 41.

However there are multiple font sanitizer warnings, see the attached ´console` log below.
It appears that this actually rendered much better, and without warnings, prior to PR #4259.


```
PDF 8407baef849645829648bab3a506d660 [1.6 TargetStream StreamEDP for Connecticut General Life Insurance Company, Inc. / TargetStream Technologies StreamEDP] (PDF.js: 1.1.415 [WebGL]) viewer.js:6708:1
downloadable font: cmap: bad glyph id offset (56170 > 56170) (font-family: "g_font_1" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: cmap: failed to parse table (font-family: "g_font_1" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: rejected by sanitizer (font-family: "g_font_1" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: font load failed (font-family: "g_font_1" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
Warning: Failed to load font ""g_font_1"": SyntaxError: An invalid or illegal string was specified pdf.js:224:5
downloadable font: cmap: bad glyph id offset (56170 > 56170) (font-family: "g_font_2" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: cmap: failed to parse table (font-family: "g_font_2" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: rejected by sanitizer (font-family: "g_font_2" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: font load failed (font-family: "g_font_2" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
Warning: Failed to load font ""g_font_2"": SyntaxError: An invalid or illegal string was specified pdf.js:224:5
downloadable font: cmap: bad glyph id offset (56170 > 56170) (font-family: "g_font_3" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: cmap: failed to parse table (font-family: "g_font_3" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: rejected by sanitizer (font-family: "g_font_3" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: font load failed (font-family: "g_font_3" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
Warning: Failed to load font ""g_font_3"": SyntaxError: An invalid or illegal string was specified pdf.js:224:5
downloadable font: cmap: bad glyph id offset (56170 > 56170) (font-family: "g_font_4" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: cmap: failed to parse table (font-family: "g_font_4" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: rejected by sanitizer (font-family: "g_font_4" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
downloadable font: font load failed (font-family: "g_font_4" style:normal weight:normal stretch:normal src index:0) source: (invalid URI) <unknown>
Warning: Failed to load font ""g_font_4"": SyntaxError: An invalid or illegal string was specified
```

**Comment by InPermutation** [comment]
Total gibberish is now back to English, as promised by @Snuffleupagus. And, indeed, the rendering is still pretty poor. At least it's readable!

**Comment by brendandahl** [comment]
Looked into this a bit. When we generate the cmap for this font the length of the table overflows the maximum allowed length of a cmap format 4 table(16bits). This appears to mainly happen because we use the CIDtoGIDMap in the PDF that has tons of char codes mapped to one glyph id (32). Format 4 is very inefficient to represent this. To fix this we could potentially use a format 12 or 13 cmap. I'm not sure how well other cmap formats are supported on the various platforms though.

**Comment by Snuffleupagus** [comment]
Since the link above appears to be dead, I'm attaching the PDF file here: [issue6397.pdf](https://github.com/mozilla/pdf.js/files/1542599/issue6397.pdf).

**Comment by timvandermeij** [comment]
Fixed by #9340.
